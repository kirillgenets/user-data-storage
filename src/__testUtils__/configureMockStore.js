import configureStore from 'redux-mock-store';

const configureMockStore = () => configureStore([]);

export default configureMockStore;
