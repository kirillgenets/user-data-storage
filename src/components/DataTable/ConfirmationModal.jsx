import React from 'react';
import PropTypes from 'prop-types';
import Modal from 'react-modal';

const ConfirmationModal = ({ onSubmit, onCancel }) => {
  return (
    <Modal isOpen onRequestClose={onCancel} className="data-table__confirmation">
      <h3 className="data-table__confirmation-text">Are you sure?</h3>
      <div className="data-table__confirmation-buttons">
        <button className="data-table__confirmation-button data-table__confirmation-button--yes" onClick={onSubmit}>
          Yes
        </button>
        <button className="data-table__confirmation-button data-table__confirmation-button--no" onClick={onCancel}>
          No
        </button>
      </div>
    </Modal>
  );
};

ConfirmationModal.propTypes = {
  onSubmit: PropTypes.func,
  onCancel: PropTypes.func,
};

export default ConfirmationModal;
