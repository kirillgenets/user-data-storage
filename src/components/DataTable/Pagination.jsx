import React from 'react';
import PropTypes from 'prop-types';
import ReactPaginate from 'react-paginate';

const Pagination = ({ onPageChange, pagesCount }) => {
  return (
    <ReactPaginate
      breakLabel="..."
      nextLabel="next >"
      onPageChange={onPageChange}
      pageRangeDisplayed={5}
      pageCount={pagesCount}
      previousLabel="< previous"
      renderOnZeroPageCount={null}
    />
  );
};

Pagination.propTypes = {
  onPageChange: PropTypes.func.isRequired,
  pagesCount: PropTypes.number,
};

Pagination.defaultProps = {
  pagesCount: 0,
};

export default Pagination;
