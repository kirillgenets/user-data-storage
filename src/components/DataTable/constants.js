import { FieldName } from '../../constants';

export const COLUMNS = [
  {
    key: FieldName.FirstName,
    name: 'First name',
  },
  {
    key: FieldName.LastName,
    name: 'Last name',
  },
  {
    key: FieldName.Email,
    name: 'E-Mail',
  },
  {
    key: FieldName.BirthDate,
    name: 'Date of birth',
  },
];

export const ITEMS_PER_PAGE = 5;
