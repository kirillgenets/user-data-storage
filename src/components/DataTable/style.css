.data-table {
  display: flex;
  flex-direction: column;
  gap: 15px;

  width: 80%;
  margin: 0 auto;
}

.data-table__search {
  width: 30%;
  margin: 0 auto;
  padding: 8px;
}

.data-table__table {
  border-collapse: collapse;
}

.data-table__row {
  border-bottom: 1px solid #ececec;
}

.data-table__row--head {
  border-color: #a1a1a1;
}

.data-table__cell {
  padding: 10px;
  text-align: left;
}

.data-table__cell--head {
  background-color: #ececec;
}

.data-table__confirmation {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  height: 100%;
}

.data-table__confirmation-buttons {
  display: flex;
  gap: 10px;
}
