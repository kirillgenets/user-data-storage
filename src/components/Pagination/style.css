.pagination {
  display: flex;
  justify-content: center;
  gap: 10px;

  list-style: none;
}

.pagination__item a {
  color: '#4f4f4f';

  cursor: pointer;
}

.pagination__item a:hover {
  color: #bababa;
}
