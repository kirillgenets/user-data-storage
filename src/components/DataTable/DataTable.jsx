import React, { useCallback, useMemo, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import Modal from 'react-modal';

import Pagination from '../Pagination';
import ConfirmationModal from './ConfirmationModal';

import { selectCustomers } from '../../selectors/customers';
import { removeCustomer } from '../../actions/customers';
import { FieldName } from '../../constants';
import { COLUMNS, ITEMS_PER_PAGE } from './constants';
import filterData from './utils/filterData';

import './style.css';

Modal.setAppElement('#root');

const DataTable = () => {
  const dispatch = useDispatch();

  const customers = useSelector(selectCustomers);

  const [itemOffset, setItemOffset] = useState(0);
  const [search, setSearch] = useState('');
  const [itemToDelete, setItemToDelete] = useState(null);
  const [isError, setIsError] = useState(false);

  const searchedCustomers = useMemo(() => filterData(customers, search), [customers, search]);

  const items = searchedCustomers.slice(itemOffset, itemOffset + ITEMS_PER_PAGE);
  const pagesCount = Math.ceil(searchedCustomers.length / ITEMS_PER_PAGE);

  const handlePageChange = useCallback(
    (evt) => {
      const newOffset = (evt.selected * ITEMS_PER_PAGE) % customers.length;
      setItemOffset(newOffset);
    },
    [customers],
  );

  const handleSearch = (evt) => {
    setSearch(evt.target.value);
    setItemOffset(0);
  };

  const handleDeleteClick = (item) => () => {
    setItemToDelete(item);
    setIsError(false);
  };

  const handleDeleteCancel = () => {
    setItemToDelete(null);
  };

  const handleDeleteSubmit = useCallback(() => {
    try {
      dispatch(removeCustomer(itemToDelete));
      setItemToDelete(null);
    } catch (err) {
      console.error(err);
      setIsError(true);
    }
  }, [dispatch, itemToDelete]);

  return customers.length > 0 ? (
    <div className="data-table">
      <input
        type="search"
        className="data-table__search"
        onChange={handleSearch}
        value={search}
        placeholder="Search..."
      />
      <table className="data-table__table">
        <thead>
          <tr className="data-table__row data-table__row--head">
            {COLUMNS.map(({ name, key }) => (
              <th className="data-table__cell data-table__cell--head" key={key}>
                {name}
              </th>
            ))}
          </tr>
        </thead>
        <tbody>
          {items.map((item) => (
            <tr className="data-table__row" key={item[FieldName.Email]}>
              {COLUMNS.map(({ key }) => (
                <td className="data-table__cell" key={key}>
                  {item[key]}
                </td>
              ))}
              <td className="data-table__cell">
                <button className="data-table__delete" onClick={handleDeleteClick(item)}>
                  Delete
                </button>
              </td>
            </tr>
          ))}
        </tbody>
      </table>
      <Pagination onPageChange={handlePageChange} pagesCount={pagesCount} />
      {itemToDelete && (
        <ConfirmationModal
          onSubmit={handleDeleteSubmit}
          onCancel={handleDeleteCancel}
          error={isError ? 'Deletion has failed due to the unexpected reason' : ''}
        />
      )}
    </div>
  ) : null;
};

export default DataTable;
