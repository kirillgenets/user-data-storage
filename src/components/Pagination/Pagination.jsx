import PropTypes from 'prop-types';
import ReactPaginate from 'react-paginate';

import './style.css';

const Pagination = ({ onPageChange, pagesCount }) => {
  return (
    <ReactPaginate
      pageClassName="pagination__item"
      previousClassName="pagination__item pagination__item--prev"
      nextClassName="pagination__item pagination__item--next"
      className="pagination"
      breakLabel="..."
      nextLabel=">"
      onPageChange={onPageChange}
      pageRangeDisplayed={5}
      pageCount={pagesCount}
      previousLabel="<"
      renderOnZeroPageCount={null}
    />
  );
};

Pagination.propTypes = {
  onPageChange: PropTypes.func.isRequired,
  pagesCount: PropTypes.number,
};

Pagination.defaultProps = {
  pagesCount: 0,
};

export default Pagination;
