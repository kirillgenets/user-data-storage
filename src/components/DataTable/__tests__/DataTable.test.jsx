import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';

import configureMockStore from '../../../__testUtils__/configureMockStore';
import { generateMockCustomers } from '../../../mock';

import DataTable from '../DataTable';
import { ITEMS_PER_PAGE } from '../constants';
import Pagination from '../../Pagination';
import { FieldName } from '../../../constants';
import { REMOVE_CUSTOMER } from '../../../actions/customers';

const mockStore = configureMockStore();

describe('DataTable', () => {
  it('should render correctly', () => {
    const store = mockStore({ customers: { items: generateMockCustomers() } });
    const component = mount(
      <Provider store={store}>
        <DataTable />
      </Provider>,
    );
    const tree = renderer.create(component);
    expect(tree).toMatchSnapshot();
    expect(component.find('.data-table')).toHaveLength(1);
  });

  it('should not be rendered without items', () => {
    const store = mockStore({ customers: { items: [] } });
    const component = mount(
      <Provider store={store}>
        <DataTable />
      </Provider>,
    );
    expect(component.find('.data-table')).toHaveLength(0);
  });

  it(`should render only ${ITEMS_PER_PAGE} items per page`, () => {
    const store = mockStore({ customers: { items: generateMockCustomers() } });
    const component = mount(
      <Provider store={store}>
        <DataTable />
      </Provider>,
    );
    expect(component.find('tbody .data-table__row')).toHaveLength(ITEMS_PER_PAGE);
  });

  it('should provide correct pages count to Pagination', () => {
    const customers = generateMockCustomers();
    const store = mockStore({ customers: { items: customers } });
    const component = mount(
      <Provider store={store}>
        <DataTable />
      </Provider>,
    );
    expect(component.find(Pagination).props().pagesCount).toBe(customers.length / ITEMS_PER_PAGE);
  });

  it('should filter customers by search', () => {
    const customers = generateMockCustomers();
    const store = mockStore({ customers: { items: customers } });
    const component = mount(
      <Provider store={store}>
        <DataTable />
      </Provider>,
    );
    expect(component.find('tbody .data-table__row')).toHaveLength(ITEMS_PER_PAGE);

    component
      .find('.data-table__search')
      .simulate('change', { target: { value: customers[customers.length - 1][FieldName.Email] } });
    expect(component.find('tbody .data-table__row')).toHaveLength(1);
  });

  it('should provide an ability to delete customers', () => {
    const customers = generateMockCustomers();
    const store = mockStore({ customers: { items: customers } });
    const component = mount(
      <Provider store={store}>
        <DataTable />
      </Provider>,
    );

    component.find('.data-table__delete').first().simulate('click');

    expect(component.find('.data-table__confirmation-button--yes')).toHaveLength(1);
    component.find('.data-table__confirmation-button--yes').simulate('click');

    const actions = store.getActions();
    expect(actions).toEqual([
      {
        type: REMOVE_CUSTOMER,
        payload: customers[0],
      },
    ]);
  });

  it('should close the confirmation when cancelling', () => {
    const customers = generateMockCustomers();
    const store = mockStore({ customers: { items: customers } });
    const component = mount(
      <Provider store={store}>
        <DataTable />
      </Provider>,
    );

    component.find('.data-table__delete').first().simulate('click');

    expect(component.find('.data-table__confirmation-button--no')).toHaveLength(1);
    component.find('.data-table__confirmation-button--no').simulate('click');
    expect(component.find('.data-table__confirmation')).toHaveLength(0);

    const actions = store.getActions();
    expect(actions).toEqual([]);
  });
});
