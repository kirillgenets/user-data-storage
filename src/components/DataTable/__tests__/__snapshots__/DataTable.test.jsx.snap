// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable should render correctly 1`] = `
<div
  className="data-table"
>
  <input
    className="data-table__search"
    onChange={[Function]}
    placeholder="Search..."
    type="search"
    value=""
  />
  <table
    className="data-table__table"
  >
    <thead>
      <tr
        className="data-table__row data-table__row--head"
      >
        <th
          className="data-table__cell data-table__cell--head"
        >
          First name
        </th>
        <th
          className="data-table__cell data-table__cell--head"
        >
          Last name
        </th>
        <th
          className="data-table__cell data-table__cell--head"
        >
          E-Mail
        </th>
        <th
          className="data-table__cell data-table__cell--head"
        >
          Date of birth
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="data-table__row"
      >
        <td
          className="data-table__cell"
        >
          Test
        </td>
        <td
          className="data-table__cell"
        >
          Test
        </td>
        <td
          className="data-table__cell"
        >
          0@example.com
        </td>
        <td
          className="data-table__cell"
        >
          1995-01-01
        </td>
        <td
          className="data-table__cell"
        >
          <button
            className="data-table__delete"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr
        className="data-table__row"
      >
        <td
          className="data-table__cell"
        >
          Test
        </td>
        <td
          className="data-table__cell"
        >
          Test
        </td>
        <td
          className="data-table__cell"
        >
          1@example.com
        </td>
        <td
          className="data-table__cell"
        >
          1995-01-01
        </td>
        <td
          className="data-table__cell"
        >
          <button
            className="data-table__delete"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr
        className="data-table__row"
      >
        <td
          className="data-table__cell"
        >
          Test
        </td>
        <td
          className="data-table__cell"
        >
          Test
        </td>
        <td
          className="data-table__cell"
        >
          2@example.com
        </td>
        <td
          className="data-table__cell"
        >
          1995-01-01
        </td>
        <td
          className="data-table__cell"
        >
          <button
            className="data-table__delete"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr
        className="data-table__row"
      >
        <td
          className="data-table__cell"
        >
          Test
        </td>
        <td
          className="data-table__cell"
        >
          Test
        </td>
        <td
          className="data-table__cell"
        >
          3@example.com
        </td>
        <td
          className="data-table__cell"
        >
          1995-01-01
        </td>
        <td
          className="data-table__cell"
        >
          <button
            className="data-table__delete"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr
        className="data-table__row"
      >
        <td
          className="data-table__cell"
        >
          Test
        </td>
        <td
          className="data-table__cell"
        >
          Test
        </td>
        <td
          className="data-table__cell"
        >
          4@example.com
        </td>
        <td
          className="data-table__cell"
        >
          1995-01-01
        </td>
        <td
          className="data-table__cell"
        >
          <button
            className="data-table__delete"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <ul
    aria-label="Pagination"
    className="pagination"
    role="navigation"
  >
    <li
      className="pagination__item pagination__item--prev disabled"
    >
      <a
        aria-disabled="true"
        aria-label="Previous page"
        className=" "
        onClick={[Function]}
        onKeyPress={[Function]}
        rel="prev"
        role="button"
        tabIndex="-1"
      >
        &lt;
      </a>
    </li>
    <li
      className="pagination__item selected"
    >
      <a
        aria-current="page"
        aria-label="Page 1 is your current page"
        onClick={[Function]}
        onKeyPress={[Function]}
        rel="canonical"
        role="button"
        tabIndex="-1"
      >
        1
      </a>
    </li>
    <li
      className="pagination__item"
    >
      <a
        aria-current={null}
        aria-label="Page 2"
        onClick={[Function]}
        onKeyPress={[Function]}
        rel="next"
        role="button"
        tabIndex="0"
      >
        2
      </a>
    </li>
    <li
      className="pagination__item"
    >
      <a
        aria-current={null}
        aria-label="Page 3"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex="0"
      >
        3
      </a>
    </li>
    <li
      className="pagination__item"
    >
      <a
        aria-current={null}
        aria-label="Page 4"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex="0"
      >
        4
      </a>
    </li>
    <li
      className="pagination__item"
    >
      <a
        aria-current={null}
        aria-label="Page 5"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex="0"
      >
        5
      </a>
    </li>
    <li
      className="break"
    >
      <a
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex="0"
      >
        ...
      </a>
    </li>
    <li
      className="pagination__item"
    >
      <a
        aria-current={null}
        aria-label="Page 18"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex="0"
      >
        18
      </a>
    </li>
    <li
      className="pagination__item"
    >
      <a
        aria-current={null}
        aria-label="Page 19"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex="0"
      >
        19
      </a>
    </li>
    <li
      className="pagination__item"
    >
      <a
        aria-current={null}
        aria-label="Page 20"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex="0"
      >
        20
      </a>
    </li>
    <li
      className="pagination__item pagination__item--next"
    >
      <a
        aria-disabled="false"
        aria-label="Next page"
        className=""
        onClick={[Function]}
        onKeyPress={[Function]}
        rel="next"
        role="button"
        tabIndex="0"
      >
        &gt;
      </a>
    </li>
  </ul>
</div>
`;
